lib/kubernetestoken/token_validator.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetestoken
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/gravitational/trace"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/strings/slices"
+)
+
+const (
+	serviceAccountGroup      = "system:serviceaccounts"
+	ServiceAccountNamePrefix = "system:serviceaccount"
+	extraDataPodNameField    = "authentication.kubernetes.io/pod-name"
+	// Kubernetes should support bound tokens on 1.20 and 1.21,
+	// but we can have an apiserver running 1.21 and kubelets running 1.19.
+	kubernetesBoundTokenSupportMinor = 22
+)
+
+type Validator struct {
+	mu sync.Mutex
+	// client is protected by mu and should only be accessed via the getProvider
+	// method.
+	client kubernetes.Interface
+}
+
+// getClient allows the lazy initialisation of the Kubernetes client
+func (v *Validator) getClient() (kubernetes.Interface, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
```

Comment:
Just to double check -- this mutex is only needed to ensure we don't try to lazy initialize the Kube `client` twice. Once the `client` is initialized, its safe for multiple goroutines to access it at the same time. IOW we can assume that whatever `kubernetes.Interface` is, its threadsafe?

Reply:
Yeah - this is based off of the designer of my token validator for GitHub etc. The mutex avoids racing on initialising the lazy-loaded client, but the client itself should be concurrency-safe. As far as I know from prior adventures, the kubeapi client is concurrency safe (as are most API clients in Go in my experience).